src/QMCWaveFunctions/SpinorSet.cpp

```diff
@@ -211,6 +211,75 @@ void SpinorSet::mw_evaluateVGLWithSpin(const RefVectorWithLeader<SPOSet>& spo_li
   }
 }
 
+void SpinorSet::mw_evaluateVGLandDetRatioGradsWithSpin(const RefVectorWithLeader<SPOSet>& spo_list,
+                                                       const RefVectorWithLeader<ParticleSet>& P_list,
+                                                       int iat,
+                                                       const std::vector<const ValueType*>& invRow_ptr_list,
+                                                       OffloadMWVGLArray& phi_vgl_v,
+                                                       std::vector<ValueType>& ratios,
+                                                       std::vector<GradType>& grads,
+                                                       std::vector<ValueType>& spingrads) const
+{
+  auto& spo_leader = spo_list.getCastedLeader<SpinorSet>();
+  auto& P_leader   = P_list.getLeader();
+  assert(this == &spo_leader);
+  assert(phi_vgl_v.size(0) == DIM_VGL);
+  assert(phi_vgl_v.size(1) == spo_list.size());
+  const size_t nw             = spo_list.size();
+  const size_t norb_requested = phi_vgl_v.size(2);
+
+  OffloadMWVGLArray up_phi_vgl_v, dn_phi_vgl_v;
+  up_phi_vgl_v.resize(DIM_VGL, nw, norb_requested);
+  dn_phi_vgl_v.resize(DIM_VGL, nw, norb_requested);
+  std::vector<ValueType> up_ratios(nw), dn_ratios(nw);
+  std::vector<GradType> up_grads(nw), dn_grads(nw);
+
+  SPOSet& up_spo_leader = *(spo_leader.spo_up);
+  SPOSet& dn_spo_leader = *(spo_leader.spo_dn);
+  RefVectorWithLeader<SPOSet> up_spo_list(up_spo_leader);
+  RefVectorWithLeader<SPOSet> dn_spo_list(dn_spo_leader);
+  up_spo_list.reserve(nw);
+  dn_spo_list.reserve(nw);
+
+  for (int iw = 0; iw < nw; iw++)
+  {
+    SpinorSet& spinor = spo_list.getCastedElement<SpinorSet>(iw);
+    up_spo_list.emplace_back(*(spinor.spo_up));
+    dn_spo_list.emplace_back(*(spinor.spo_dn));
+  }
+
+  up_spo_leader.mw_evaluateVGLandDetRatioGrads(up_spo_list, P_list, iat, invRow_ptr_list, up_phi_vgl_v, up_ratios,
+                                               up_grads);
+  dn_spo_leader.mw_evaluateVGLandDetRatioGrads(dn_spo_list, P_list, iat, invRow_ptr_list, dn_phi_vgl_v, dn_ratios,
+                                               dn_grads);
+
+  for (int iw = 0; iw < nw; iw++)
```

Comment:
We will need to make this computation on GPU. But can be done in a separate PR.

Reply:
added the updateFrom calls and some comments noting that we need to revisit this eventually